File: #69_Sqrt/Solution.java
public class Solution {
    public int mySqrt(int x) {
    	if (x < 0) throw new IllegalArgumentException();
        if (x == 0) return 0;
        if (x == 1) return 1;
        return (int) sqrt(x, 1, x - 1);
    }

    private long sqrt(int target, long lo, long hi) {
    	if (lo == hi) return lo;
    	if (lo == hi - 1) {
    		if (hi * hi == target) return hi;
    		if (lo * lo == target) return lo;
    		if (target > hi * hi) return hi;
    		return lo;
    	}
    	long sum = lo + hi;
    	long mid = sum / 2;
    	/* careful for overflow */
    	if (mid == target / mid) return mid;
    	/* can't skip mid here */
    	if (mid < target / mid) return sqrt(target, mid, hi);
    	if (mid > target / mid) return sqrt(target, lo, mid);
    	return -1;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	System.out.println(sol.mySqrt(8));
    }
}

File: #69_Sqrt/Solution.py
'''
Binary search
time: O(log(n))
'''

class Solution:
    # @param {integer} x
    # @return {integer}
    def mySqrt(self, x):
        if x == 0 or x == 1:
        	return x
        minVal = 1
        maxVal = x
        while minVal <= maxVal:
        	mid = minVal + (maxVal - minVal) / 2
        	if mid * mid == x:
        		return mid
        	elif mid * mid > x:
        		maxVal = mid - 1
        	else:
        		minVal = mid + 1
        return min(maxVal, minVal)

sol = Solution()
print sol.mySqrt(10)